### DISCUSSION TITLE
저장시 client 객체를 가져오는 법

### DISCUSSION THREAD

**Discussion by peterkangdev** [naverz/zepeto-studio-kor#309] [5. Q&A]
저장하는 예시를 보면 다음과 같은 코드를 보는데요..

const playerStorage:DataStorage = client.loadDataStorage();

이때 client 객체는 어디서 가져오나요?
멀티 플레이 예시를 보면 ClientStarter 라는 클래스에 이벤트를 받아오는 로직에 있던데.

저는 다른 곳에서 저장을 하고 싶어요.. (ClientStarter 에서 한다고 해도 어느 부분에 저장 로직을 넣어야 하는지도 모르겠고.)
client 객체를 가져오는 방법을 모르겠습니다.

**Comment by hyeonjin-jo** [comment]
안녕하세요.

해당 코드는 서버쪽에서 저장소를 로드할 때 사용되는 부분입니다.

https://studio.zepeto.me/guides/multiplay-game-datastorage

위의 문서를 참고 부탁드리고 서버 쪽의 OnJoin함수의 매개변수로 SandboxPlayer타입의 client객체가 존재합니다.


```typescript
import { DataStorage } from "ZEPETO.Multiplay.DataStorage";

async onJoin(client: SandboxPlayer) {        
        const playerStorage:DataStorage = client.loadDataStorage();
        let playerLevel = await playerStorage.get("level") as number;

        if(playerLevel == null)
            playerLevel = 1;

        await playerStorage.set("level", playerLevel);
}  
```

**Comment by peterkangdev** [comment]
저장하는 방법을 알고 싶습니다. (hyeonjin-jo 님의 코드를 몰라서 하는 얘기가 아니라)

가령 멀티플레이에서 사용자가 먹은 아이템의 갯수를 저장하고 싶습니다.
아이템을 먹을때 마다 서버에 저장해야 하는데
저장하는 스크립트는 hyeonjin-jo 님의 예에서 처럼 onJoin() 은 아닐것입니다. 
그리고 저장하는 로직이 위에서 처럼 loadDataStorage() 함수를 써야 한다면 
SandboxPlayer 객체를 얻어와야 하는데, 이 객체는 어떻게 얻는지 궁금한 것입니다.
아님 저장하는 방법이 따로 있는지 궁금하네요..

말이 장황해서 이해 가셨을지 모르겠네요..!!

**Reply by hyeonjin-jo** [reply]
OnJoin에서 획득 가능한 Storage를 현재 코드처럼 지역변수가 아닌 멤버 변수로 만들어서 사용하는 것을 추천드립니다.

클라에서 특정 이벤트일 때 서버 쪽에서 해당 멤버 변수를 이용해서 저장하는 코드를 활용해 보시기 바랍니다.

playerStorage.set 함수는 자주 호출되면 성능상 문제가 발생하므로 월드 자체가 종료되었을 때 저장하는 방법을 권장드립니다.

```typescript
storageMap:Map<string,DataStorage> = new Map<string, DataStorage>();

async onJoin(client: SandboxPlayer) {        
        const playerStorage:DataStorage = client.loadDataStorage();
        this.storageMap.set(client.sessionId,playerStorage);
}

```
